File: backend/src/datafusion/schemas/system.py
"""
System Mode API schemas.

Schemas for the surveillance operator dashboard, case management,
and decision submission endpoints.
"""
import enum
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from datafusion.schemas.domains import DomainType
from datafusion.schemas.risk import CorrelationAlert, RecommendedAction, RiskAssessment


# === Enums ===


class AlertType(str, enum.Enum):
    """Types of system alerts shown to operators."""

    QUOTA_WARNING = "quota_warning"
    REVIEW_PENDING = "review_pending"
    DIRECTIVE_UPDATE = "directive_update"
    COMMENDATION = "commendation"


class AlertUrgency(str, enum.Enum):
    """Urgency level for system alerts."""

    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class ComplianceTrend(str, enum.Enum):
    """Trend direction for compliance metrics."""

    IMPROVING = "improving"
    STABLE = "stable"
    DECLINING = "declining"


# === Dashboard Schemas ===


class DirectiveRead(BaseModel):
    """Directive information for display."""

    id: UUID
    directive_key: str
    week_number: int
    title: str
    description: str
    internal_memo: str | None = Field(
        description="Revealed after directive completion"
    )
    required_domains: list[str]
    flag_quota: int
    time_limit_hours: int | None
    moral_weight: int
    content_rating: str


class DailyMetrics(BaseModel):
    """Operator's daily performance metrics."""

    flags_today: int = Field(description="Flags submitted today")
    quota: int = Field(description="Required flags for current directive")
    average_decision_time: float = Field(description="Average seconds per decision")
    compliance_trend: ComplianceTrend = Field(description="Compliance direction")


class SystemAlert(BaseModel):
    """Alert shown on operator dashboard."""

    alert_type: AlertType
    message: str
    urgency: AlertUrgency


class OperatorStatusRead(BaseModel):
    """Operator status for dashboard display."""

    operator_id: UUID
    operator_code: str
    status: str
    compliance_score: float
    current_quota_progress: str
    total_flags_submitted: int
    total_reviews_completed: int
    hesitation_incidents: int
    warnings: list[str] = Field(default_factory=list)


class SystemDashboard(BaseModel):
    """Complete dashboard data for System Mode."""

    operator: OperatorStatusRead
    directive: DirectiveRead | None
    metrics: DailyMetrics
    alerts: list[SystemAlert]
    pending_cases: int


# === Case Schemas ===


class CaseOverview(BaseModel):
    """Brief overview of a case for the case list."""

    npc_id: UUID
    name: str
    age: int
    risk_score: int
    risk_level: str
    primary_concern: str = Field(description="Top contributing factor")
    flagged_messages: int = Field(description="Number of flagged messages")
    time_in_queue: str = Field(description="How long since case was generated")


class MessageRead(BaseModel):
    """Message for display in citizen file."""

    id: UUID
    timestamp: datetime
    recipient_name: str
    recipient_relationship: str | None
    content: str
    is_flagged: bool
    flag_reasons: list[str]
    sentiment: float
    detected_keywords: list[str]


class CitizenFlagRead(BaseModel):
    """Previous flag record for display."""

    id: UUID
    flag_type: str
    created_at: datetime
    justification: str
    outcome: str


class IdentityRead(BaseModel):
    """Citizen identity information."""

    id: UUID
    first_name: str
    last_name: str
    date_of_birth: str
    age: int
    ssn: str
    street_address: str
    city: str
    state: str
    zip_code: str


class FullCitizenFile(BaseModel):
    """Complete citizen file with all available data."""

    identity: IdentityRead
    risk_assessment: RiskAssessment
    domains: dict[str, dict] = Field(description="Domain data by domain type")
    messages: list[MessageRead]
    correlation_alerts: list[CorrelationAlert]
    recommended_actions: list[RecommendedAction]
    flag_history: list[CitizenFlagRead]


# === Decision Submission Schemas ===


class FlagSubmission(BaseModel):
    """Request to submit a flag against a citizen."""

    operator_id: UUID
    citizen_id: UUID
    flag_type: str = Field(description="monitoring, restriction, intervention, detention")
    contributing_factors: list[str] = Field(description="Factor keys from risk assessment")
    justification: str = Field(min_length=10, description="Operator's stated reason")
    decision_time_seconds: float = Field(description="Time spent on decision")


class MetricsDelta(BaseModel):
    """Changes to operator metrics after a decision."""

    compliance_change: float
    quota_progress: str
    hesitation_flagged: bool


class FlagResult(BaseModel):
    """Result of submitting a flag."""

    flag_id: UUID
    citizen_id: UUID  # Added for cinematic transitions
    citizen_name: str
    flag_type: str
    immediate_outcome: str
    quota_progress: str
    compliance_score: float
    warnings: list[str]
    metrics_delta: MetricsDelta


class NoActionSubmission(BaseModel):
    """Request to skip flagging a citizen (requires justification)."""

    operator_id: UUID
    citizen_id: UUID
    justification: str = Field(
        min_length=20, description="REQUIRED: Explain why no action taken"
    )
    decision_time_seconds: float


class NoActionResult(BaseModel):
    """Result of submitting a no-action decision."""

    logged: bool
    compliance_impact: float = Field(description="Negative impact on compliance")
    cumulative_no_actions: int
    warning: str | None = None


class FlagSummary(BaseModel):
    """Summary of a flag for history view."""

    flag_id: UUID
    citizen_name: str
    flag_type: str
    created_at: datetime
    outcome: str
    one_line_summary: str


# === Session Start ===


class SystemStartRequest(BaseModel):
    """Request to start System Mode session."""

    session_id: UUID


class SystemStartResponse(BaseModel):
    """Response after starting System Mode."""

    operator_id: UUID
    operator_code: str
    status: str
    compliance_score: float
    first_directive: DirectiveRead

File: frontend/src/types/system.ts
/**
 * System Mode TypeScript types.
 * Mirrors backend schemas for the surveillance operator interface.
 */

// === Enums ===

export type OperatorStatusType = 'active' | 'under_review' | 'suspended' | 'terminated';
export type FlagType = 'monitoring' | 'restriction' | 'intervention' | 'detention';
export type FlagOutcome = 'pending' | 'actioned' | 'rejected' | 'appealed';
export type RiskLevel = 'low' | 'moderate' | 'elevated' | 'high' | 'severe';
export type AlertType = 'quota_warning' | 'review_pending' | 'directive_update' | 'commendation';
export type AlertUrgency = 'info' | 'warning' | 'critical';
export type ComplianceTrend = 'improving' | 'stable' | 'declining';
export type ActionType = 'increase_monitoring' | 'travel_restriction' | 'employer_notification' | 'intervention' | 'detention';
export type ActionUrgency = 'routine' | 'priority' | 'immediate';
export type EndingType = 'compliant_operator' | 'reluctant_operator' | 'suspended_operator' | 'resistance_path';

// === Operator & Dashboard ===

export interface OperatorStatus {
  operator_id: string;
  operator_code: string;
  status: OperatorStatusType;
  compliance_score: number;
  current_quota_progress: string;
  total_flags_submitted: number;
  total_reviews_completed: number;
  hesitation_incidents: number;
  warnings: string[];
}

export interface DirectiveRead {
  id: string;
  directive_key: string;
  week_number: number;
  title: string;
  description: string;
  internal_memo: string | null;
  required_domains: string[];
  flag_quota: number;
  time_limit_hours: number | null;
  moral_weight: number;
  content_rating: string;
}

export interface DailyMetrics {
  flags_today: number;
  quota: number;
  average_decision_time: number;
  compliance_trend: ComplianceTrend;
}

export interface SystemAlert {
  alert_type: AlertType;
  message: string;
  urgency: AlertUrgency;
}

export interface SystemDashboard {
  operator: OperatorStatus;
  directive: DirectiveRead | null;
  metrics: DailyMetrics;
  alerts: SystemAlert[];
  pending_cases: number;
}

export interface SystemDashboardWithCases {
  dashboard: SystemDashboard;
  cases: CaseOverview[];
}

// === Risk Assessment ===

export interface ContributingFactor {
  factor_key: string;
  factor_name: string;
  weight: number;
  evidence: string;
  domain_source: string;
}

export interface CorrelationAlert {
  alert_type: string;
  description: string;
  confidence: number;
  domains_involved: string[];
}

export interface RecommendedAction {
  action_type: ActionType;
  justification: string;
  urgency: ActionUrgency;
}

export interface RiskAssessment {
  npc_id: string;
  risk_score: number;
  risk_level: RiskLevel;
  contributing_factors: ContributingFactor[];
  correlation_alerts: CorrelationAlert[];
  recommended_actions: RecommendedAction[];
  last_updated: string;
}

// === Cases ===

export interface CaseOverview {
  npc_id: string;
  name: string;
  age: number;
  risk_score: number;
  risk_level: RiskLevel;
  primary_concern: string;
  flagged_messages: number;
  time_in_queue: string;
}

export interface MessageRead {
  id: string;
  timestamp: string;
  recipient_name: string;
  recipient_relationship: string | null;
  content: string;
  is_flagged: boolean;
  flag_reasons: string[];
  sentiment: number;
  detected_keywords: string[];
}

export interface CitizenFlagRead {
  id: string;
  flag_type: FlagType;
  created_at: string;
  justification: string;
  outcome: FlagOutcome;
}

export interface IdentityRead {
  id: string;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  age: number;
  ssn: string;
  street_address: string;
  city: string;
  state: string;
  zip_code: string;
}

export interface FullCitizenFile {
  identity: IdentityRead;
  risk_assessment: RiskAssessment;
  domains: Record<string, Record<string, unknown>>;
  messages: MessageRead[];
  correlation_alerts: CorrelationAlert[];
  recommended_actions: RecommendedAction[];
  flag_history: CitizenFlagRead[];
}

// === Decisions ===

export interface FlagSubmission {
  operator_id: string;
  citizen_id: string;
  flag_type: FlagType;
  contributing_factors: string[];
  justification: string;
  decision_time_seconds: number;
}

export interface MetricsDelta {
  compliance_change: number;
  quota_progress: string;
  hesitation_flagged: boolean;
}

export interface FlagResult {
  flag_id: string;
  citizen_id: string;  // Added for cinematic transitions
  citizen_name: string;
  flag_type: FlagType;
  immediate_outcome: string;
  quota_progress: string;
  compliance_score: number;
  warnings: string[];
  metrics_delta: MetricsDelta;
}

export interface NoActionSubmission {
  operator_id: string;
  citizen_id: string;
  justification: string;
  decision_time_seconds: number;
}

export interface NoActionResult {
  logged: boolean;
  compliance_impact: number;
  cumulative_no_actions: number;
  warning: string | null;
}

export interface FlagSummary {
  flag_id: string;
  citizen_name: string;
  flag_type: FlagType;
  created_at: string;
  outcome: FlagOutcome;
  one_line_summary: string;
}

// === Outcomes ===

export interface CitizenOutcome {
  flag_id: string;
  citizen_id: string;  // Added for cinematic transitions
  citizen_name: string;
  time_skip: string;
  status: string;
  narrative: string;
  statistics: Record<string, unknown>;
}

export interface CitizenOutcomeSummary {
  citizen_name: string;
  flag_type: FlagType;
  final_status: string;
  one_line_summary: string;
}

// === Operator Assessment ===

export interface OperatorContributingFactor {
  factor_key: string;
  factor_name: string;
  weight: number;
  evidence: string;
}

export interface OperatorRiskAssessment {
  operator_code: string;
  risk_score: number;
  risk_level: string;
  contributing_factors: OperatorContributingFactor[];
  recommended_action: string;
  assessment_date: string;
}

// === Ending ===

export interface RealWorldExample {
  name: string;
  country: string;
  year: string;
  description: string;
}

export interface RealWorldParallel {
  title: string;
  description: string;
  examples: RealWorldExample[];
  call_to_action: string;
}

export interface EducationalLink {
  title: string;
  url: string;
  description: string;
}

export interface EndingStatistics {
  total_citizens_flagged: number;
  lives_disrupted: number;
  families_separated: number;
  detentions_ordered: number;
  jobs_destroyed: number;
  your_compliance_score: number;
  your_risk_score: number | null;
  total_decisions: number;
  hesitation_incidents: number;
}

export interface EndingResult {
  ending_type: EndingType;
  title: string;
  narrative: string;
  statistics: EndingStatistics;
  citizens_flagged: CitizenOutcomeSummary[];
  operator_final_status: string;
  real_world_content: RealWorldParallel;
  educational_links: EducationalLink[];
}

// === Session ===

export interface SystemStartResponse {
  operator_id: string;
  operator_code: string;
  status: OperatorStatusType;
  compliance_score: number;
  first_directive: DirectiveRead;
}

// === Cinematic Transitions ===

export interface CinematicData {
  citizenId: string;
  citizenName: string;
  timeSkip: string;  // "immediate", "1_month", "6_months", "1_year"
  narrative: string;
  status: string;
  map_x: number;
  map_y: number;
}
